src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -0,0 +1,248 @@
+                 << bundle->BundleId().first << bundle->BundleId().second;
+  auto status = lease_client->RequestResourceReserve(
+      *bundle, [this, node_id, bundle, node, callback](
+                   const Status &status, const rpc::RequestResourceReserveReply &reply) {
+        // TODO(AlisaWu): Add placement group cancel.
+        auto iter = node_to_bundles_when_leasing_.find(node_id);
+        if (iter != node_to_bundles_when_leasing_.end()) {
+          auto bundle_iter = iter->second.find(bundle->BundleId());
+          RAY_CHECK(bundle_iter != iter->second.end());
+          if (status.ok()) {
+            if (reply.success()) {
+              RAY_LOG(DEBUG) << "Finished leasing resource from " << node_id
+                             << " for bundle " << bundle->BundleId().first
+                             << bundle->BundleId().second;
+            } else {
+              RAY_LOG(DEBUG) << "Failed leasing resource from " << node_id
+                             << " for bundle " << bundle->BundleId().first
+                             << bundle->BundleId().second;
+            }
+            // Remove the actor from the leasing map as the reply is returned from the
+            // remote node.
+            iter->second.erase(bundle_iter);
+            if (iter->second.empty()) {
+              node_to_bundles_when_leasing_.erase(iter);
+            }
+          }
+          callback(status, reply);
+        }
+      });
+  if (!status.ok()) {
+    rpc::RequestResourceReserveReply reply;
+    reply.set_success(false);
+    callback(status, reply);
+  }
+}
+
+void GcsPlacementGroupScheduler::CancelResourceReserve(
+    std::shared_ptr<BundleSpecification> bundle_spec,
+    std::shared_ptr<ray::rpc::GcsNodeInfo> node) {
+  RAY_CHECK(node);
+
+  auto node_id = ClientID::FromBinary(node->node_id());
+  RAY_LOG(DEBUG) << "Start returning resource for node " << node_id << " for bundle "
+                 << bundle_spec->BundleId().first << bundle_spec->BundleId().second;
+
+  rpc::Address remote_address;
+  remote_address.set_raylet_id(node->node_id());
+  remote_address.set_ip_address(node->node_manager_address());
+  remote_address.set_port(node->node_manager_port());
+  auto return_client = GetOrConnectLeaseClient(remote_address);
+  auto status = return_client->CancelResourceReserve(
+      *bundle_spec,
+      [this, bundle_spec, node](const Status &status,
+                                const rpc::CancelResourceReserveReply &reply) {
+        if (!status.ok()) {
+          return_timer_.expires_from_now(boost::posix_time::milliseconds(5));
+          return_timer_.async_wait(
+              [this, bundle_spec, node](const boost::system::error_code &error) {
+                if (error == boost::system::errc::operation_canceled) {
+                  return;
+                } else {
+                  CancelResourceReserve(bundle_spec, node);
+                }
+              });
+        }
+      });
+  if (!status.ok()) {
```

Comment:
I believe we might still fire the async error callback even if the status here is failure. Which might double cancel the resource reservation.

@stephanie-wang is that right?

Reply:
Yes, you should just ignore the returned status.